### PR TITLE
feat(drivers): add scrypt driver

### PR DESCRIPTION
**Pull request by RomainLanz**
Hey there! 👋🏻 

This PR adds the native `scrypt` hash algorithm to hash passwords.

I would recommend that we use it as default when installing a new application.

### CODE REVIEW

**Comment by targos** [comment]
Why not argon2 as the default?

**Comment by RomainLanz** [comment]
> Why not argon2 as the default?

That would be the best, but it means we have to pre-install the `argon2` package on the userland code for any installation of Adonis. Meaning also compiling its code (since it is not JS based).

`scrypt` is built directly in Node. No need to pre-install anything, and you can use it without changing the footprint of the user application

Currently, if the user installs the `auth` module (which uses the hash API by default), the user gets an error because argon is not installed.

**Comment by thetutlage** [comment]
Can you open another PR for the docs to mention that one should be using `argon2` in production. 

It can be added to https://docs.adonisjs.com/guides/security/hashing and the https://docs.adonisjs.com/guides/deployment docs

**Comment by RomainLanz** [comment]
Done! Also, PRing the `core` module to change the default driver.

Before merging, should we rename options `keylen` to `keyLength` and `maxmem` to `maxMemory`?
I used [the same naming you can find in Node](https://nodejs.org/api/crypto.html#cryptoscryptpassword-salt-keylen-options-callback), but using complete naming may be more accessible.

**Comment by RomainLanz** [comment]
After a discussion with @rault-a, we can remove `@kdf/salt` and `tsse` by using native code.
I will update the PR and migrate the other drivers.

**Comment by thetutlage** [comment]
Looking forward to it :)

> Before merging, should we rename options keylen to keyLength and maxmem to maxMemory?

Wouldn't this be a breaking change?

**Comment by RomainLanz** [comment]
> Wouldn't this be a breaking change?

Those options are only used in `scrypt`, so it will not break anything.

**Comment by RomainLanz** [comment]
Should be good!

File: src/Drivers/Scrypt.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    // Identifier Validation
+    if (!this.ids.includes(deserializedHash.id)) {
+      return Promise.reject(
+        new TypeError(`Incompatible ${deserializedHash.id} identifier found in the hash`)
+      )
+    }
+
+    // Parameters Existence Validation
+    if (typeof deserializedHash.params !== 'object') {
+      return Promise.reject(new TypeError('The param section cannot be empty'))
```
Comment (thetutlage) [inline src/Drivers/Scrypt.ts]: Can't we just thrown the exception vs creating another promise?

File: test/scrypt.spec.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+import { test } from '@japa/runner'
+import phc from '@phc/format'
+import { Scrypt } from '../src/Drivers/Scrypt'
+
+test.group('Scrypt', () => {
```
Comment (thetutlage) [inline test/scrypt.spec.ts:15]: We will need more tests covering all the failure cases as well

**Review by thetutlage — CHANGES_REQUESTED** [review]

**Comment by RomainLanz** [comment]
Done!

**Comment by thetutlage** [comment]
Looks great. Just for my personal knowledge. Are they are any recommendations guides on how to best use Scrypt? And what should be the default config for `blockSize` or `parallelization`?

**Comment by RomainLanz** [comment]
We could increase the key length to `128` bits instead of `64` to be more resilient on brute force attacks, but I believe the default value proposed by Node.js is more than enough.

Everything is a trade-off regarding the server you will run the code on. Default values should run anywhere.

**Comment by thetutlage** [comment]
Looks great 👍

**Pull request merged**